File: ScreenGuardian-Mobile/lib/services/local_store.dart
/// Local Store - File-based JSON persistence layer
/// Data format matches ScreenGuardian Desktop for cross-platform sync

import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:path_provider/path_provider.dart';
import 'package:path/path.dart' as p;
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';
import '../constants.dart';
import '../models/types.dart';
import '../utils/time_utils.dart';

class LocalStore {
  static LocalStore? _instance;
  late Directory _dataDir;
  late DeviceInfo _deviceInfo;
  late AppConfig _config;
  late AppState _state;

  LocalStore._();

  static Future<LocalStore> getInstance() async {
    if (_instance == null) {
      _instance = LocalStore._();
      await _instance!._init();
    }
    return _instance!;
  }

  Future<void> _init() async {
    final appDir = await getApplicationDocumentsDirectory();
    _dataDir = Directory(p.join(appDir.path, 'screenguardian'));
    if (!await _dataDir.exists()) {
      await _dataDir.create(recursive: true);
    }
    // Ensure subdirectories exist
    for (final subdir in ['sessions', 'summaries', 'plans', 'tracking/llm-ranking', 'devices']) {
      final dir = Directory(p.join(_dataDir.path, subdir));
      if (!await dir.exists()) {
        await dir.create(recursive: true);
      }
    }
    _loadOrCreateDeviceInfo();
    _loadOrCreateConfig();
    _loadOrCreateState();
    _performCrashRecovery();
  }

  String get dataPath => _dataDir.path;

  // ============================================================
  // File I/O helpers
  // ============================================================

  Future<Map<String, dynamic>?> _readJson(String relativePath) async {
    final file = File(p.join(_dataDir.path, relativePath));
    if (!await file.exists()) return null;
    try {
      final content = await file.readAsString();
      return jsonDecode(content) as Map<String, dynamic>;
    } catch (e) {
      return null;
    }
  }

  Future<List<dynamic>> _readJsonList(String relativePath) async {
    final file = File(p.join(_dataDir.path, relativePath));
    if (!await file.exists()) return [];
    try {
      final content = await file.readAsString();
      return jsonDecode(content) as List<dynamic>;
    } catch (e) {
      return [];
    }
  }

  Future<void> _writeJson(String relativePath, dynamic data) async {
    final file = File(p.join(_dataDir.path, relativePath));
    await file.parent.create(recursive: true);
    await file.writeAsString(const JsonEncoder.withIndent('  ').convert(data));
  }

  // ============================================================
  // Device Info
  // ============================================================

  void _loadOrCreateDeviceInfo() {
    final json = _readJsonSync('device.json');

    if (json != null) {
      _deviceInfo = DeviceInfo.fromJson(json);
    } else {
      final platform = _detectPlatform();
      final prefix = platform == PlatformType.iphone ? 'ios' : 'android';
      _deviceInfo = DeviceInfo(
        deviceId: 'dev-$prefix-${const Uuid().v4().substring(0, 8)}',
        deviceName: _defaultDeviceName(platform),
        platform: platform,
        registeredAt: DateTime.now(),
        appVersion: appVersion,
      );
      _writeJsonSync('device.json', _deviceInfo.toJson());
    }
    _deviceInfo.lastActiveAt = DateTime.now();
    _deviceInfo.appVersion = appVersion;
    _writeJsonSync('device.json', _deviceInfo.toJson());

    // Persist device ID to SharedPreferences (async, non-blocking)
    _persistDeviceId(_deviceInfo.deviceId);
  }

  Future<void> _persistDeviceId(String deviceId) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final persistedId = prefs.getString('sg_device_id');
      if (persistedId == null) {
        await prefs.setString('sg_device_id', deviceId);
      } else if (persistedId != deviceId) {
        // Use persisted ID (survives reinstalls)
        _deviceInfo.deviceId = persistedId;
        _writeJsonSync('device.json', _deviceInfo.toJson());
      }
    } catch (e) {
      print('[LocalStore] SharedPreferences not available: $e');
    }
  }

  static PlatformType _detectPlatform() {
    if (Platform.isAndroid) return PlatformType.androidPhone;
    if (Platform.isIOS) return PlatformType.iphone;
    return PlatformType.iphone;
  }

  static String _defaultDeviceName(PlatformType platform) {
    switch (platform) {
      case PlatformType.androidPhone:
        return 'Android Device';
      case PlatformType.iphone:
        return 'iOS Device';
      default:
        return 'Mobile Device';
    }
  }

  DeviceInfo get deviceInfo => _deviceInfo;
  String get deviceId => _deviceInfo.deviceId;
  String get deviceName => _config.deviceName ?? _deviceInfo.deviceName;

  // ============================================================
  // AppConfig
  // ============================================================

  void _loadOrCreateConfig() {
    final json = _readJsonSync('config.json');
    if (json != null) {
      _config = AppConfig.fromJson(json);
    } else {
      _config = AppConfig(
        updatedAt: DateTime.now(),
        updatedBy: deviceId,
      );
      _writeJsonSync('config.json', _config.toJson());
    }
  }

  AppConfig get config => _config;

  Future<AppConfig> updateConfig(Map<String, dynamic> patch) async {
    final json = _config.toJson();
    json.addAll(patch);
    json['updatedAt'] = DateTime.now().toIso8601String();
    json['updatedBy'] = deviceId;
    _config = AppConfig.fromJson(json);
    await _writeJson('config.json', _config.toJson());
    return _config;
  }

  // ============================================================
  // AppState
  // ============================================================

  void _loadOrCreateState() {
    final json = _readJsonSync('state.json');
    if (json != null) {
      _state = AppState.fromJson(json);
    } else {
      _state = AppState();
      _writeJsonSync('state.json', _state.toJson());
    }
  }

  AppState get state => _state;

  Future<void> updateState(Map<String, dynamic> patch) async {
    final json = _state.toJson();
    json.addAll(patch);
    _state = AppState.fromJson(json);
    await _writeJson('state.json', _state.toJson());
  }

  // ============================================================
  // Sessions
  // ============================================================

  String _monthKey(String date) => date.substring(0, 7);

  Future<List<ScreenSession>> loadSessions(String monthKey) async {
    final list = await _readJsonList('sessions/$monthKey.json');
    final sessions = <ScreenSession>[];
    for (final e in list) {
      try {
        sessions.add(ScreenSession.fromJson(e as Map<String, dynamic>));
      } catch (err) {
        print('[LocalStore] Skipping corrupted session: $err');
      }
    }
    return sessions;
  }

  List<ScreenSession> _loadSessionsSync(String monthKey) {
    final list = _readJsonListSync('sessions/$monthKey.json');
    return list.map((e) => ScreenSession.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<ScreenSession> createSession({DateTime? startTime}) async {
    final now = startTime ?? DateTime.now();
    final session = ScreenSession(
      id: const Uuid().v4(),
      deviceId: deviceId,
      deviceName: deviceName,
      platform: _deviceInfo.platform,
      startTime: now,
      date: formatDate(now),
      createdAt: now,
      updatedAt: now,
    );
    final mk = _monthKey(session.date);
    final sessions = await loadSessions(mk);
    sessions.add(session);
    await _writeJson('sessions/$mk.json', sessions.map((s) => s.toJson()).toList());
    await updateState({'currentSessionId': session.id});
    return session;
  }

  Future<ScreenSession?> endSession(String sessionId, StopReason reason, {DateTime? endTime}) async {
    final now = endTime ?? DateTime.now();
    final mk = _monthKey(formatDate(now));
    final sessions = await loadSessions(mk);

    var session = sessions.where((s) => s.id == sessionId).firstOrNull;
    if (session == null) {
      // Try previous month
      final prevDate = DateTime(now.year, now.month - 1, 1);
      final prevMk = _monthKey(formatDate(prevDate));
      final prevSessions = await loadSessions(prevMk);
      session = prevSessions.where((s) => s.id == sessionId).firstOrNull;
      if (session != null) {
        return _endSessionInList(prevSessions, prevMk, sessionId, reason, now);
      }
      return null;
    }
    return _endSessionInList(sessions, mk, sessionId, reason, now);
  }

  Future<ScreenSession?> _endSessionInList(
    List<ScreenSession> sessions,
    String monthKey,
    String sessionId,
    StopReason reason,
    DateTime endTime,
  ) async {
    final index = sessions.indexWhere((s) => s.id == sessionId);
    if (index == -1) return null;

    final session = sessions[index];
    session.endTime = endTime;
    session.durationSeconds = endTime.difference(session.startTime).inSeconds.clamp(0, 999999);
    session.stopReason = reason;
    session.updatedAt = DateTime.now();
    session.version++;

    await _writeJson('sessions/$monthKey.json', sessions.map((s) => s.toJson()).toList());
    await updateDailySummary(session.date);
    await updateState({'currentSessionId': null});
    return session;
  }

  Future<int> getTodayTotalSeconds() async {
    final today = todayDate();
    final mk = _monthKey(today);
    final sessions = await loadSessions(mk);
    int total = 0;
    for (final s in sessions) {
      if (s.date == today && s.durationSeconds != null) {
        total += s.durationSeconds!;
      }
    }
    // Add current session elapsed
    if (_state.currentSessionId != null) {
      final current = sessions.where((s) => s.id == _state.currentSessionId).firstOrNull;
      if (current != null && current.date == today) {
        total += DateTime.now().difference(current.startTime).inSeconds.clamp(0, 999999);
      }
    }
    return total;
  }

  // ============================================================
  // DailySummary
  // ============================================================

  Future<List<DailySummary>> loadSummaries(String monthKey) async {
    final list = await _readJsonList('summaries/$monthKey.json');
    return list.map((e) => DailySummary.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<DailySummary> updateDailySummary(String date) async {
    final mk = _monthKey(date);
    final sessions = await loadSessions(mk);
    final daySessions = sessions.where((s) => s.date == date && s.endTime != null).toList();
    final summaries = await loadSummaries(mk);

    final totalSeconds = daySessions.fold(0, (sum, s) => sum + (s.durationSeconds ?? 0));
    final devices = daySessions.map((s) => s.deviceId).toSet().toList();
    final sessionIds = daySessions.map((s) => s.id).toList();
    final startTimes = daySessions.map((s) => s.startTime).toList()..sort();
    final endTimes = daySessions.where((s) => s.endTime != null).map((s) => s.endTime!).toList()..sort();

    final newSummary = DailySummary(
      date: date,
      totalSeconds: totalSeconds,
      sessionCount: daySessions.length,
      devices: devices,
      firstSessionStart: startTimes.isNotEmpty ? startTimes.first.toIso8601String() : null,
      lastSessionEnd: endTimes.isNotEmpty ? endTimes.last.toIso8601String() : null,
      sessionIds: sessionIds,
      updatedAt: DateTime.now(),
    );

    final existingIndex = summaries.indexWhere((s) => s.date == date);
    if (existingIndex >= 0) {
      summaries[existingIndex] = newSummary;
    } else {
      summaries.add(newSummary);
    }
    await _writeJson('summaries/$mk.json', summaries.map((s) => s.toJson()).toList());
    return newSummary;
  }

  Future<DailySummary?> getDailySummary(String date) async {
    final mk = _monthKey(date);
    final summaries = await loadSummaries(mk);
    return summaries.where((s) => s.date == date).firstOrNull;
  }

  // ============================================================
  // Query helpers for reports
  // ============================================================

  Future<List<ScreenSession>> querySessions({
    required String startDate,
    required String endDate,
    String? deviceId,
    StopReason? stopReason,
  }) async {
    final startMonth = startDate.substring(0, 7);
    final endMonth = endDate.substring(0, 7);
    final months = _getMonthRange(startMonth, endMonth);
    final results = <ScreenSession>[];

    for (final mk in months) {
      final sessions = await loadSessions(mk);
      for (final s in sessions) {
        if (s.date.compareTo(startDate) < 0 || s.date.compareTo(endDate) > 0) continue;
        if (deviceId != null && s.deviceId != deviceId) continue;
        if (stopReason != null && s.stopReason != stopReason) continue;
        results.add(s);
      }
    }
    results.sort((a, b) => a.startTime.compareTo(b.startTime));
    return results;
  }

  Future<List<DailySummary>> querySummaries({
    required String startDate,
    required String endDate,
  }) async {
    final startMonth = startDate.substring(0, 7);
    final endMonth = endDate.substring(0, 7);
    final months = _getMonthRange(startMonth, endMonth);
    final results = <DailySummary>[];

    for (final mk in months) {
      final summaries = await loadSummaries(mk);
      for (final s in summaries) {
        if (s.date.compareTo(startDate) >= 0 && s.date.compareTo(endDate) <= 0) {
          results.add(s);
        }
      }
    }
    results.sort((a, b) => a.date.compareTo(b.date));
    return results;
  }

  List<String> _getMonthRange(String startMonth, String endMonth) {
    final months = <String>[];
    var current = startMonth;
    while (current.compareTo(endMonth) <= 0) {
      months.add(current);
      final parts = current.split('-').map(int.parse).toList();
      final nextMonth = parts[1] == 12 ? 1 : parts[1] + 1;
      final nextYear = parts[1] == 12 ? parts[0] + 1 : parts[0];
      current = '${nextYear}-${nextMonth.toString().padLeft(2, '0')}';
    }
    return months;
  }

  // ============================================================
  // Crash Recovery
  // ============================================================

  Future<void> _performCrashRecovery() async {
    if (_state.currentSessionId != null) {
      final session = await endSession(_state.currentSessionId!, StopReason.shutdown);
      if (session != null) {
        print('[CrashRecovery] Closed orphaned session ${session.id}, duration: ${session.durationSeconds}s');
      }
      await updateState({'currentSessionId': null});
    }
  }

  // ============================================================
  // Weekly Plans
  // ============================================================

  Future<List<WeeklyPlan>> loadWeeklyPlans() async {
    final list = await _readJsonList('plans/weekly.json');
    return list.map((e) => WeeklyPlan.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<void> saveWeeklyPlan(WeeklyPlan plan) async {
    final plans = await loadWeeklyPlans();
    final index = plans.indexWhere((p) => p.weekStart == plan.weekStart);
    if (index >= 0) {
      plans[index] = plan;
    } else {
      plans.add(plan);
    }
    await _writeJson('plans/weekly.json', plans.map((p) => p.toJson()).toList());
  }

  Future<WeeklyPlan?> getWeeklyPlan(String weekStart) async {
    final plans = await loadWeeklyPlans();
    return plans.where((p) => p.weekStart == weekStart).firstOrNull;
  }

  // ============================================================
  // Sync helpers (expose raw read/write for SyncService)
  // ============================================================

  Future<Map<String, dynamic>?> readRawJson(String relativePath) => _readJson(relativePath);
  Future<void> writeRawJson(String relativePath, dynamic data) => _writeJson(relativePath, data);

  // ============================================================
  // Synchronous helpers (for init-time use only)
  // ============================================================

  Map<String, dynamic>? _readJsonSync(String relativePath) {
    try {
      final file = File(p.join(_dataDir.path, relativePath));
      if (!file.existsSync()) return null;
      final content = file.readAsStringSync();
      return jsonDecode(content) as Map<String, dynamic>;
    } catch (e) {
      return null;
    }
  }

  List<dynamic> _readJsonListSync(String relativePath) {
    try {
      final file = File(p.join(_dataDir.path, relativePath));
      if (!file.existsSync()) return [];
      final content = file.readAsStringSync();
      return jsonDecode(content) as List<dynamic>;
    } catch (e) {
      return [];
    }
  }

  void _writeJsonSync(String relativePath, dynamic data) {
    try {
      final file = File(p.join(_dataDir.path, relativePath));
      file.parent.createSync(recursive: true);
      file.writeAsStringSync(const JsonEncoder.withIndent('  ').convert(data));
    } catch (e) {
      print('[LocalStore] Sync write error: $e');
    }
  }
}

/// App state - tracks current session, timer state, etc.
class AppState {
  String? currentSessionId;
  int? eyeRestTimerRemainingMs;
  int? postureTimerRemainingMs;
  String lastScreenState;
  bool meetingMode;
  String? lastWeeklySummaryWeek;
  bool overtimeAlertedToday;
  String? lastOvertimeReminderAt;

  AppState({
    this.currentSessionId,
    this.eyeRestTimerRemainingMs,
    this.postureTimerRemainingMs,
    this.lastScreenState = 'active',
    this.meetingMode = false,
    this.lastWeeklySummaryWeek,
    this.overtimeAlertedToday = false,
    this.lastOvertimeReminderAt,
  });

  Map<String, dynamic> toJson() => {
    'currentSessionId': currentSessionId,
    'eyeRestTimerRemainingMs': eyeRestTimerRemainingMs,
    'postureTimerRemainingMs': postureTimerRemainingMs,
    'lastScreenState': lastScreenState,
    'meetingMode': meetingMode,
    'lastWeeklySummaryWeek': lastWeeklySummaryWeek,
    'overtimeAlertedToday': overtimeAlertedToday,
    'lastOvertimeReminderAt': lastOvertimeReminderAt,
  };

  factory AppState.fromJson(Map<String, dynamic> json) => AppState(
    currentSessionId: json['currentSessionId'],
    eyeRestTimerRemainingMs: json['eyeRestTimerRemainingMs'],
    postureTimerRemainingMs: json['postureTimerRemainingMs'],
    lastScreenState: json['lastScreenState'] ?? 'active',
    meetingMode: json['meetingMode'] ?? false,
    lastWeeklySummaryWeek: json['lastWeeklySummaryWeek'],
    overtimeAlertedToday: json['overtimeAlertedToday'] ?? false,
    lastOvertimeReminderAt: json['lastOvertimeReminderAt'],
  );
}

File: ScreenGuardian-Desktop/src/services/local-store.ts
/**
 * Local Store - File-based JSON persistence layer
 * Data format matches ScreenGuardian Mobile for cross-platform sync
 */

import * as fs from 'fs';
import * as path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { app } from 'electron';
import {
  ScreenSession, DailySummary, WeeklyPlan, DeviceInfo, AppConfig, AppState,
  PlatformType, StopReason
} from './types';
import { todayDate, formatDate } from '../utils/time-utils';

const APP_VERSION: string = require('../../../package.json').version;

export class LocalStore {
  private dataDir: string;
  private _deviceInfo!: DeviceInfo;
  private _config!: AppConfig;
  private _state!: AppState;

  constructor() {
    // Store data next to the executable (as noted in DESIGN.md)
    const exeDir = path.dirname(process.execPath);
    this.dataDir = path.join(exeDir, 'screenguardian');
    this.ensureDirectories();
    this.loadOrCreateDeviceInfo();
    this.loadOrCreateConfig();
    this.loadOrCreateState();
    this.performCrashRecovery();
  }

  get deviceId(): string { return this._deviceInfo.deviceId; }
  get deviceName(): string { return this._config.deviceName || this._deviceInfo.deviceName; }
  get config(): AppConfig { return this._config; }
  get state(): AppState { return this._state; }
  get deviceInfo(): DeviceInfo { return this._deviceInfo; }
  get dataPath(): string { return this.dataDir; }

  // ============================================================
  // Directory setup
  // ============================================================

  private ensureDirectories(): void {
    const dirs = [
      this.dataDir,
      path.join(this.dataDir, 'sessions'),
      path.join(this.dataDir, 'summaries'),
      path.join(this.dataDir, 'plans'),
      path.join(this.dataDir, 'tracking', 'llm-ranking'),
      path.join(this.dataDir, 'devices'),
    ];
    for (const dir of dirs) {
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
    }
  }

  // ============================================================
  // File I/O helpers
  // ============================================================

  private readJson(relativePath: string): any | null {
    try {
      const filePath = path.join(this.dataDir, relativePath);
      if (!fs.existsSync(filePath)) return null;
      return JSON.parse(fs.readFileSync(filePath, 'utf-8'));
    } catch {
      return null;
    }
  }

  private readJsonList(relativePath: string): any[] {
    try {
      const filePath = path.join(this.dataDir, relativePath);
      if (!fs.existsSync(filePath)) return [];
      return JSON.parse(fs.readFileSync(filePath, 'utf-8')) as any[];
    } catch {
      return [];
    }
  }

  private writeJson(relativePath: string, data: any): void {
    const filePath = path.join(this.dataDir, relativePath);
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    // Atomic write: write to temp file, then rename
    const tmpPath = `${filePath}.tmp`;
    fs.writeFileSync(tmpPath, JSON.stringify(data, null, 2), 'utf-8');
    fs.renameSync(tmpPath, filePath);
  }

  // ============================================================
  // Device Info
  // ============================================================

  private loadOrCreateDeviceInfo(): void {
    const json = this.readJson('device.json');
    if (json) {
      this._deviceInfo = json as DeviceInfo;
    } else {
      const platform = process.platform === 'win32' ? PlatformType.Windows : PlatformType.MacOS;
      const prefix = process.platform === 'win32' ? 'win' : 'mac';
      const os = require('os');
      this._deviceInfo = {
        deviceId: `dev-${prefix}-${uuidv4().substring(0, 8)}`,
        deviceName: `${os.userInfo().username}的${process.platform === 'win32' ? 'Windows电脑' : 'MacBook'}`,
        platform,
        registeredAt: new Date().toISOString(),
        lastSyncAt: null,
        lastActiveAt: new Date().toISOString(),
        appVersion: APP_VERSION,
      };
      this.writeJson('device.json', this._deviceInfo);
    }
    this._deviceInfo.lastActiveAt = new Date().toISOString();
    this._deviceInfo.appVersion = APP_VERSION;
    this.writeJson('device.json', this._deviceInfo);
  }

  // ============================================================
  // AppConfig
  // ============================================================

  private loadOrCreateConfig(): void {
    const json = this.readJson('config.json');
    if (json) {
      this._config = json as AppConfig;
    } else {
      this._config = {
        language: 'system',
        postureIntervalMinutes: 30,
        trackingTargets: ['llm_ranking'],
        meetingMode: false,
        syncFolderPath: null,
        deviceName: null,
        eyeRestEnabled: true,
        postureEnabled: true,
        overtimeEnabled: true,
        updatedAt: new Date().toISOString(),
        updatedBy: this.deviceId,
      };
      this.writeJson('config.json', this._config);
    }
  }

  async updateConfig(patch: Partial<AppConfig>): Promise<AppConfig> {
    Object.assign(this._config, patch, {
      updatedAt: new Date().toISOString(),
      updatedBy: this.deviceId,
    });
    this.writeJson('config.json', this._config);
    return this._config;
  }

  // ============================================================
  // AppState
  // ============================================================

  private loadOrCreateState(): void {
    const json = this.readJson('state.json');
    if (json) {
      this._state = json as AppState;
    } else {
      this._state = {
        currentSessionId: null,
        eyeRestTimerRemainingMs: null,
        postureTimerRemainingMs: null,
        lastScreenState: 'active',
        meetingMode: false,
        lastWeeklySummaryWeek: null,
        overtimeAlertedToday: false,
        lastOvertimeReminderAt: null,
      };
      this.writeJson('state.json', this._state);
    }
  }

  async updateState(patch: Partial<AppState>): Promise<void> {
    Object.assign(this._state, patch);
    this.writeJson('state.json', this._state);
  }

  // ============================================================
  // Sessions
  // ============================================================

  private monthKey(date: string): string {
    return date.substring(0, 7);
  }

  loadSessions(monthKey: string): ScreenSession[] {
    return this.readJsonList(`sessions/${monthKey}.json`) as ScreenSession[];
  }

  async createSession(startTime?: Date): Promise<ScreenSession> {
    const now = startTime || new Date();
    const session: ScreenSession = {
      id: uuidv4(),
      deviceId: this.deviceId,
      deviceName: this.deviceName,
      platform: this._deviceInfo.platform,
      startTime: now.toISOString(),
      endTime: null,
      durationSeconds: null,
      stopReason: null,
      date: formatDate(now),
      createdAt: now.toISOString(),
      updatedAt: now.toISOString(),
      version: 1,
    };
    const mk = this.monthKey(session.date);
    const sessions = this.loadSessions(mk);
    sessions.push(session);
    this.writeJson(`sessions/${mk}.json`, sessions);
    await this.updateState({ currentSessionId: session.id });
    return session;
  }

  async endSession(sessionId: string, reason: StopReason, endTime?: Date): Promise<ScreenSession | null> {
    const now = endTime || new Date();
    const mk = this.monthKey(formatDate(now));
    let sessions = this.loadSessions(mk);
    let session = sessions.find(s => s.id === sessionId);

    if (!session) {
      // Try previous month
      const prevDate = new Date(now.getFullYear(), now.getMonth() - 1, 1);
      const prevMk = this.monthKey(formatDate(prevDate));
      const prevSessions = this.loadSessions(prevMk);
      session = prevSessions.find(s => s.id === sessionId);
      if (session) {
        return this.endSessionInList(prevSessions, prevMk, sessionId, reason, now);
      }
      return null;
    }
    return this.endSessionInList(sessions, mk, sessionId, reason, now);
  }

  private endSessionInList(
    sessions: ScreenSession[], monthKey: string,
    sessionId: string, reason: StopReason, endTime: Date
  ): ScreenSession | null {
    const index = sessions.findIndex(s => s.id === sessionId);
    if (index === -1) return null;

    const session = sessions[index];
    session.endTime = endTime.toISOString();
    session.durationSeconds = Math.max(0,
      Math.floor((endTime.getTime() - new Date(session.startTime).getTime()) / 1000));
    session.stopReason = reason;
    session.updatedAt = new Date().toISOString();
    session.version++;

    this.writeJson(`sessions/${monthKey}.json`, sessions);
    this.updateDailySummary(session.date);
    this.updateState({ currentSessionId: null });
    return session;
  }

  async getTodayTotalSeconds(): Promise<number> {
    const today = todayDate();
    const mk = this.monthKey(today);
    const sessions = this.loadSessions(mk);

    // Collect all time intervals for today
    const intervals: [Date, Date][] = [];
    for (const s of sessions) {
      if (s.date === today && s.endTime) {
        intervals.push([new Date(s.startTime), new Date(s.endTime)]);
      }
    }
    // Include current active session
    if (this._state.currentSessionId) {
      const current = sessions.find(s => s.id === this._state.currentSessionId);
      if (current && current.date === today) {
        intervals.push([new Date(current.startTime), new Date()]);
      }
    }

    return this.computeDedupedSeconds(intervals);
  }

  // ============================================================
  // DailySummary
  // ============================================================

  loadSummaries(monthKey: string): DailySummary[] {
    return this.readJsonList(`summaries/${monthKey}.json`) as DailySummary[];
  }

  updateDailySummary(date: string): DailySummary {
    const mk = this.monthKey(date);
    const sessions = this.loadSessions(mk);
    const daySessions = sessions.filter(s => s.date === date && s.endTime != null);
    const summaries = this.loadSummaries(mk);

    // Deduplicate overlapping time intervals across devices
    const intervals: [Date, Date][] = daySessions
      .filter(s => s.endTime)
      .map(s => [new Date(s.startTime), new Date(s.endTime!)] as [Date, Date]);
    const totalSeconds = this.computeDedupedSeconds(intervals);
    const devices = [...new Set(daySessions.map(s => s.deviceId))];
    const sessionIds = daySessions.map(s => s.id);
    const startTimes = daySessions.map(s => new Date(s.startTime)).sort((a, b) => a.getTime() - b.getTime());
    const endTimes = daySessions
      .filter(s => s.endTime)
      .map(s => new Date(s.endTime!))
      .sort((a, b) => a.getTime() - b.getTime());

    const newSummary: DailySummary = {
      date,
      totalSeconds,
      sessionCount: daySessions.length,
      devices,
      firstSessionStart: startTimes.length > 0 ? startTimes[0].toISOString() : null,
      lastSessionEnd: endTimes.length > 0 ? endTimes[endTimes.length - 1].toISOString() : null,
      sessionIds,
      updatedAt: new Date().toISOString(),
    };

    const existingIndex = summaries.findIndex(s => s.date === date);
    if (existingIndex >= 0) {
      summaries[existingIndex] = newSummary;
    } else {
      summaries.push(newSummary);
    }
    this.writeJson(`summaries/${mk}.json`, summaries);
    return newSummary;
  }

  getDailySummary(date: string): DailySummary | null {
    const mk = this.monthKey(date);
    const summaries = this.loadSummaries(mk);
    return summaries.find(s => s.date === date) || null;
  }

  /**
   * Compute total unique seconds from a list of time intervals.
   * Merges overlapping intervals so time is never double-counted.
   * Handles the case where multiple devices record screen time
   * during the same period (e.g., phone + laptop both active 9:00-10:00).
   */
  computeDedupedSeconds(intervals: [Date, Date][]): number {
    if (intervals.length === 0) return 0;

    // Sort by start time
    intervals.sort((a, b) => a[0].getTime() - b[0].getTime());

    // Merge overlapping intervals
    const merged: [Date, Date][] = [[intervals[0][0], intervals[0][1]]];
    for (let i = 1; i < intervals.length; i++) {
      const current = intervals[i];
      const last = merged[merged.length - 1];

      if (current[0].getTime() <= last[1].getTime()) {
        // Overlapping or adjacent — extend
        if (current[1].getTime() > last[1].getTime()) {
          merged[merged.length - 1] = [last[0], current[1]];
        }
      } else {
        // No overlap — new interval
        merged.push([current[0], current[1]]);
      }
    }

    // Sum merged intervals
    let total = 0;
    for (const [start, end] of merged) {
      total += Math.max(0, Math.floor((end.getTime() - start.getTime()) / 1000));
    }
    return total;
  }

  /** Deduped seconds for a date range (for reports) */
  getDedupedTotalSeconds(startDate: string, endDate: string): number {
    const sessions = this.querySessions({ startDate, endDate });
    const intervals: [Date, Date][] = sessions
      .filter(s => s.endTime)
      .map(s => [new Date(s.startTime), new Date(s.endTime!)] as [Date, Date]);
    return this.computeDedupedSeconds(intervals);
  }

  // ============================================================
  // Query helpers
  // ============================================================

  querySessions(opts: { startDate: string; endDate: string; deviceId?: string; stopReason?: StopReason }): ScreenSession[] {
    const months = this.getMonthRange(opts.startDate.substring(0, 7), opts.endDate.substring(0, 7));
    const results: ScreenSession[] = [];
    for (const mk of months) {
      const sessions = this.loadSessions(mk);
      for (const s of sessions) {
        if (s.date < opts.startDate || s.date > opts.endDate) continue;
        if (opts.deviceId && s.deviceId !== opts.deviceId) continue;
        if (opts.stopReason && s.stopReason !== opts.stopReason) continue;
        results.push(s);
      }
    }
    results.sort((a, b) => a.startTime.localeCompare(b.startTime));
    return results;
  }

  querySummaries(opts: { startDate: string; endDate: string }): DailySummary[] {
    const months = this.getMonthRange(opts.startDate.substring(0, 7), opts.endDate.substring(0, 7));
    const results: DailySummary[] = [];
    for (const mk of months) {
      const summaries = this.loadSummaries(mk);
      for (const s of summaries) {
        if (s.date >= opts.startDate && s.date <= opts.endDate) {
          results.push(s);
        }
      }
    }
    results.sort((a, b) => a.date.localeCompare(b.date));
    return results;
  }

  private getMonthRange(startMonth: string, endMonth: string): string[] {
    const months: string[] = [];
    let current = startMonth;
    while (current <= endMonth) {
      months.push(current);
      const [y, m] = current.split('-').map(Number);
      const nextM = m === 12 ? 1 : m + 1;
      const nextY = m === 12 ? y + 1 : y;
      current = `${nextY}-${String(nextM).padStart(2, '0')}`;
    }
    return months;
  }

  // ============================================================
  // Crash Recovery
  // ============================================================

  private performCrashRecovery(): void {
    if (this._state.currentSessionId) {
      const session = this.endSession(this._state.currentSessionId, StopReason.Shutdown);
      if (session) {
        console.log(`[CrashRecovery] Closed orphaned session ${session.id}, duration: ${session.durationSeconds}s`);
      }
      this.updateState({ currentSessionId: null });
    }
  }

  // ============================================================
  // Weekly Plans
  // ============================================================

  loadWeeklyPlans(): WeeklyPlan[] {
    return this.readJsonList('plans/weekly.json') as WeeklyPlan[];
  }

  saveWeeklyPlan(plan: WeeklyPlan): void {
    const plans = this.loadWeeklyPlans();
    const index = plans.findIndex(p => p.weekStart === plan.weekStart);
    if (index >= 0) {
      plans[index] = plan;
    } else {
      plans.push(plan);
    }
    this.writeJson('plans/weekly.json', plans);
  }

  getWeeklyPlan(weekStart: string): WeeklyPlan | null {
    const plans = this.loadWeeklyPlans();
    return plans.find(p => p.weekStart === weekStart) || null;
  }

  // ============================================================
  // Raw I/O for sync
  // ============================================================

  readRawJson(relativePath: string): any | null {
    return this.readJson(relativePath);
  }

  writeRawJson(relativePath: string, data: any): void {
    this.writeJson(relativePath, data);
  }
}
